README.md
# declarative-db
![NpmVersion](https://img.shields.io/npm/v/declarative-db.svg?style=flat-square)
![npm](https://img.shields.io/npm/dt/declarative-db.svg?style=flat-square)
![NodeVersion](https://img.shields.io/node/v/declarative-db.svg?style=flat-square)
![GitHub](https://img.shields.io/github/license/loarca/declarative-db.svg?style=flat-square)
![GitHub stars](https://img.shields.io/github/stars/loarca/declarative-db.svg?style=flat-square&label=Stars)
![GitHub issues](https://img.shields.io/github/issues/loarca/declarative-db.svg?style=flat-square)
![GitHub closed issues](https://img.shields.io/github/issues-closed/loarca/declarative-db.svg?style=flat-square)

An easy-to-use declarative json-based database for Node

## Installation
```sh
npm install declarative-db
```

## Usage
```js
const path = require('path');
const declare = require('declarative-db');

// Declare json-based database
declare({
  filename: path.join(__dirname, './db.json'),
}).then(db => {
  // Use it as you would normally use an object (or array)
  db.users = [{
    username: 'loarca',
  }];

  // It will automatically save to disk asynchronously when appropriate
});
````

Since it's Promise based you can also use async-await
```js
const path = require('path');
const declare = require('declarative-db');

(async () => {
  // Declare json-based database
  let db = await declare({
    filename: path.join(__dirname, './db.json'),
  });

  // Use it as you would normally use an object (or array)
  db.users = [{
    username: 'loarca',
  }];

  // It will automatically save to disk asynchronously when appropriate
})();
```

You should `declare()` only once per file
and share declared objects/arrays across all modules,
as every declared object/array is... exactly that... an object/array in memory
that automatically saves to disk when appropriate

## Options
```js
{
  filename: path.join(__dirname, './db.json'),
  initialState: {},
  compression: 0,
}
```

### filename
File to store data

### initialState (default = {})
When no database file exists `initialState` is used,
it may be an object or array

### compression (default = 0)
Compression level ranging from `0` to `9`

Level `0` outputs database to readable json
```json
{
  "users": [
    {
      "username": "loarca"
    }
  ]
}
```
Levels from `1` to `9` match DEFLATE compression levels

## To do
- Documentation
- Extensive testing
- *~~Smart scheduled saving~~*
- Protect scheduled saving from unexpected process termination
- When `filename` option is not present, manage database in memory
- Force saving when desired


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.3] - 2019-11-26
### Fixed
- Refactor to use Typescript.

## [1.0.2] - 2018-08-16
### Fixed
- `initialState` now works as intended, it was always the initial state
whether there was already a file or not (issue #7).

## [1.0.1] - 2018-08-13
### Added
- Mutex protection to saving to disk.
- Smart scheduled saving.

## 1.0.0 - 2018-08-13
### Added
- Stateful database.
- `compression` option.

[Unreleased]: https://github.com/loarca/declarative-db/compare/v1.0.3...HEAD
[1.0.3]: https://github.com/loarca/declarative-db/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/loarca/declarative-db/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/loarca/declarative-db/compare/v1.0.0...v1.0.1


src/index.ts
import { loadState, isObject } from './utils'
import { proxyState } from './proxy'
import { State, DeclarativeDbOptions } from './types'
export { State, DeclarativeDbOptions }

/**
 * Create a declarative database.
 * @param {DeclarativeDbOptions} options - Options for database creation.
 */
export default async (options: DeclarativeDbOptions): Promise<State> => {
  // Deconstruct options
  let { filename, compression, initialState } = options

  // Make sure initialState is an object (or array)
  initialState = isObject(initialState) ? initialState : {} as State

  // Make sure compression option is valid
  if (typeof compression === 'undefined') compression = 0
  if (typeof compression !== 'number' || compression < 0 || compression > 9) {
    console.warn('[declarative-db]', '`compression` option must be a number between 0 and 9, using 0 by default.')
    compression = 0
  }

  // Load state from disk, create a new one if doesn't exist
  const state = await loadState(filename, initialState)

  // Create proxied object
  return proxyState(state, filename, compression)
}


src/proxy.ts
import { Mutex } from 'async-mutex'
import { isObject, scheduleSaving } from './utils'
import { State, StateMetadata } from './types'

export const proxyState = (state: State, filename: string, compression: number): State => {
  // Metadata
  const metadata: StateMetadata = {
    state,
    filename,
    compression,
    mutex: new Mutex(),
    savingID: null,
    unsavedChanges: 0
  }

  // Proxy handler
  const handler: ProxyHandler<State> = {
    // When getting a property
    get (target: State, property: PropertyKey, receiver: any): any {
      // Get value normally
      const propertyValue = Reflect.get(target, property, receiver)

      // If it's proxiable, make it a proxy in order to watch nested changes
      if (isObject(propertyValue)) {
        return new Proxy(propertyValue, handler)
      } else {
        return propertyValue
      }
    },

    // When a property is set
    set (target: State, property: PropertyKey, value: any, receiver: any): boolean {
      scheduleSaving(metadata)
      return Reflect.set(target, property, value, receiver)
    },

    // When a property is deleted
    deleteProperty (target: State, property: PropertyKey): boolean {
      scheduleSaving(metadata)
      return Reflect.deleteProperty(target, property)
    }
  }

  return new Proxy<State>(state, handler)
}


src/state.js
const fs = require('fs');
const zlib = require('zlib');
const { promisify } = require('util');
const Mutex = require('async-mutex').Mutex;

const MAX_UNSAVED_CHANGES = 1000;
const MAX_UNSAVED_TIMEOUT = 60 * 1000;

class State {
  constructor({state, filename, compression}) {
    this.state = state;
    this.filename = filename;
    this.compression = compression;

    this.mutex = new Mutex();
    this.savingID = null;
    this.unsavedChanges = 0;
  }

  setState(state) {
    // Save state in memory
    this.state = state;

    // Schedule saving to disk
    this.scheduleSaving();
  }

  scheduleSaving() {
    this.unsavedChanges++;

    // If unsaved changes limit is exceeded, save state immediately
    if (this.unsavedChanges >= MAX_UNSAVED_CHANGES) {
      clearTimeout(this.savingID);
      this.savingID = null;
      this.unsavedChanges = 0;
      this.saveToDisk();
    } else {
      // Schedule saving to disk
      clearTimeout(this.savingID);
      this.savingID = setTimeout(() => {
        this.savingID = null;
        this.unsavedChanges = 0;
        this.saveToDisk();
      }, MAX_UNSAVED_TIMEOUT);
    }
  }

  async saveToDisk() {
    let unlock = () => null;

    try {
      // Lock mutex to protect saving to disk
      unlock = await this.mutex.acquire();

      // Get string data
      let data = JSON.stringify(this.state, null, this.compression ? null : 2);

      // Compress data if specified
      if (this.compression) data = await promisify(zlib.deflate)(data, {
        level: this.compression,
      });

      // Save state to disk
      await promisify(fs.writeFile)(this.filename, data);
    } catch (err) {
      console.error('[declarative-db]', err);
    } finally {
      // Unlock mutex
      unlock();
    }
  }
}

module.exports = async ({initialState, filename, compression}) => {
  // Make sure compression option is valid
  if (typeof compression !== 'number' || compression < 0 || compression > 9)
    throw new Error(
      'declarative-db: compression option must be a number between 0 and 9.'
    );

  // Determine initial state
  let state = initialState;
  // If the file specified by filename exists, load contents
  try {
    let rawState = await promisify(fs.readFile)(filename);

    // Determine if it's compressed
    try {
      // If it is, uncompress it
      rawState = await promisify(zlib.unzip)(rawState);
    } catch (err) {
      // If it is not, do nothing
    }

    state = JSON.parse(rawState);
  } catch (err) {
    // If the file specified by filename doesn't exist, use basic state = {}
    // Otherwise, rethrow
    if (err.code !== 'ENOENT')
      throw err;
  }

  // Generate new State object
  return new State({state, filename, compression});
};


src/types.ts
import { Mutex } from 'async-mutex'

/**
 * Type for declarative-db database.
 * It can be an object or array.
 * @typedef {object|Array} State
 */
export type State = Record<PropertyKey, any> | Array<any>

/**
 * Options for creating a declarative database.
 * @typedef {object} DeclarativeDbOptions
 * @property {string} filename - The filename where the database file will be saved.
 * @property {number} [compression=0] - Compression level ranging from 0 to 9 (both inclusive).
 * @property {State} [initialState={}] - Initial state that would be used if filename doesn't exist.
 */
export interface DeclarativeDbOptions {
  filename: string
  compression?: 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
  initialState?: State
}

/**
 * Metadata for a single proxied database object.
 * @typedef {object} StateMetadata
 * @private
 */
export interface StateMetadata {
  state: State
  filename: string
  compression: number
  mutex: Mutex
  savingID: NodeJS.Timeout | null
  unsavedChanges: number
}


src/utils.ts
import * as fs from 'fs'
import * as zlib from 'zlib'
import { promisify } from 'util'
import { State, StateMetadata } from './types'

const MAX_UNSAVED_CHANGES = 1000
const MAX_UNSAVED_TIMEOUT = 60 * 1000

export const isObject = (obj: any): obj is State => {
  const type = typeof obj
  return type === 'function' || (type === 'object' && !!obj)
}

const saveToDisk = async (metadata: StateMetadata): Promise<void> => {
  let unlock = (): void => { }

  try {
    // Lock mutex to protect saving to disk
    unlock = await metadata.mutex.acquire()

    // Get string data
    let data: string | Buffer = JSON.stringify(metadata.state, null, metadata.compression ? undefined : 2)

    // Compress data if specified
    if (metadata.compression) {
      data = await promisify<zlib.InputType, zlib.ZlibOptions, Buffer>(zlib.deflate)(data, {
        level: metadata.compression
      })
    }

    // Save state to disk
    await promisify(fs.writeFile)(metadata.filename, data)
  } catch (err) {
    console.error('[declarative-db]', err)
  } finally {
    // Unlock mutex
    unlock()
  }
}

export const scheduleSaving = (metadata: StateMetadata): void => {
  metadata.unsavedChanges++

  // If unsaved changes limit is exceeded, save state immediately
  if (metadata.unsavedChanges >= MAX_UNSAVED_CHANGES) {
    if (metadata.savingID) clearTimeout(metadata.savingID)
    metadata.savingID = null
    metadata.unsavedChanges = 0
    saveToDisk(metadata)
  } else {
    // Schedule saving to disk
    if (metadata.savingID) clearTimeout(metadata.savingID)
    metadata.savingID = setTimeout(() => {
      metadata.savingID = null
      metadata.unsavedChanges = 0
      saveToDisk(metadata)
    }, MAX_UNSAVED_TIMEOUT)
  }
}

export const loadState = async (filename: string, initialState: State): Promise<State> => {
  // Determine initial state
  let state = initialState

  // If the file specified by filename exists, load contents
  try {
    let rawState = await promisify(fs.readFile)(filename)

    // Determine if it's compressed
    try {
      // If it is, uncompress it
      rawState = await promisify(zlib.unzip)(rawState) as Buffer
    } catch (err) {
      // If it is not, do nothing
    }

    state = JSON.parse(rawState.toString()) as State
  } catch (err) {
    // If the file specified by filename doesn't exist, use initialState
    // Otherwise, rethrow
    if (err.code !== 'ENOENT') { throw err }
  }

  // Return unproxied state
  return state
}
